### ISSUE TITLE
feature: configurable single source of truth for keymaps

### ISSUE THREAD

**Issue by pathetic-lynx** [LazyVim/LazyVim#2193]
### Did you check the docs?

- [X] I have read all the LazyVim docs

### Is your feature request related to a problem? Please describe.

I - as several others i realise after skimming reddit and the issues - have some problems changing LazyVims default keymaps. I am not saying it is not possible. I am just saying it requires knowledge of how lazyloading works and a multitude of changes in different files (overriding plugin options, changes in my own keymap.lua) along with a lot of copy pasting of anonymous functions that i cannot just map to.

My usecase in particular was that I wanted to change the "prefix" for code actions from `<leader>c` to `<leader>C`.

From my experience this is a problem that some (most?) of the neovim distributions have (i have used LunarVim and NvChad before). Let me know if this has been adressed in another neovim distribution in some way.

### Describe the solution you'd like

I would like a definition file covering all keymaps like this:
```
{
  "PrefixCode" = "<leader>c",
  "KeymapMason" = "%PrefixCode%m",
  "KeymapFormatInjected" = "%PrefixCode%F",
  ...
}
```

So if i want to change the defaults, i can just override in my file:
```
{
  "KeymapFormatInjected" = "gFF", -- just change one particular keycombo
  "PrefixCode" = "<leader>C", -- change the prefix and everything that depends on it
}
```

at startup the two dictionaries are merged and populated (replace placeholders by actual values until the dict does not change anymore). The result is the available throughout all the plugin initializations, which only refer to the ids of keymaps looking them up in the merged dict.

This would make changing defaults a lot more straighforward and would require no knowledge about when and how plugins are loaded and keymaps are set.

### Describe alternatives you've considered

Manually remapping each individual keycombo that involves `<leader>c` which not only is a major pita but also will probably break on some future change where a LazyVim extra introduces a new keycombo.

### Additional context

If you do not have any considerations about performance impact or other caveats i havent thought about, i would love to work on this if it has a chance of getting merged.

**Comment by abeldekat** [comment]
Hello @pathetic-lynx,

Reading your excellent feature request, I discern two use-cases.

**First use case**: Change one or more individual keys

> My usecase in particular was that I wanted to change the "prefix" for code actions from <leader>c to <leader>C.

This is only one key, `<leader>ca`.

Changing lsp keys is explained in the LazyVim docs:
[lsp keymaps](https://www.lazyvim.org/configuration/keymaps#lsp-keymaps)
[customizing lsp keymaps](https://www.lazyvim.org/plugins/lsp#%EF%B8%8F-customizing-lsp-keymaps)

The knowledge obtained by changing the key is very valuable.
After all, LazyVim is intended to make it easy to customize and extend *your* config.

**Second use case**: Change a complete top level menu

> Manually remapping each individual keycombo that involves <leader>c which not only is a major pita but also will probably break on some future change where a LazyVim extra introduces a new keycombo.

I agree that in this case the amount of work is considerable.
Especially given the intent of the user: To use the corresponding `leader` key for something else.

I wrote a plugin implementing the second use case: [lazyvim-menu-addon](https://github.com/abeldekat/lazyvim-menu-addon)
The reason for writing the plugin is explained [here](https://github.com/abeldekat/lazyvim-menu-addon#builtin-to-lazyvim)

The plugin is announced in the discussions section: [show and tell](https://github.com/LazyVim/LazyVim/discussions/2323)

I would be very interested in your feedback!

@folke, what do you think?

**Comment by pathetic-lynx** [comment]
Hi @abeldekat,

first of all thanks for writing a plugin! <3 This was really keeping me from using Lazyvim. I was working on sth myself but it would've probably taken me much longer.

> First use case: Change one or more individual keys

I did non mean code action but rather "anything that starts with `<leader>c`". But the two usecases exist nonetheless and I share your opinion that using the already well-documented way of changing keybinds is preferred here.

There still is an advantage when rebinding is done separately from plugin configuration, namely that when changing a keybind you sometimes also have to add plugin options unrelated to the keybind to your customization config. This leads to potentially incompatible plugin config when someone updates the plugin and changes the defaults or invalidates the snippet that was copied to change the keybind.
Somebody going the way to change individual keybinds for a plugin that person might already have changed the default config or would not mind fixing the config when it breaks (which i would assume does not happen that often) so i believe this is not an issue.

I totally agree with you on the cons of my approach, it would require everybody to be aware of the definition file and how to use it.

I can see a middle way where the keybinds (again a `ExposedFunctionality -> Keybind` map) are a separate part of the plugin configuration thus avoiding the problems of incompatible configs.

I think for now I'll use your plugin and report back :)

**Comment by abeldekat** [comment]
Thanks, @pathetic-lynx!

I think I don't fully understand what you mean when you mention "incompatible config".
Basically, whenever a user starts customizing `LazyVim`, there is always a chance that the base spec being customized changes. At its core, in my opinion, `LazyVim` is just a bunch of `lazy.nvim` specs. That's part of its power: The user can customize almost everything without limitations.

As a consequence, that user might need to inspect incoming updates for changes that might affect those customizations.

> Somebody going the way to change individual keybinds for a plugin that person might already have changed the default config or would not mind fixing the config when it breaks (which i would assume does not happen that often) so i believe this is not an issue.

Looking at my [config](https://github.com/abeldekat/nvim) I think I can say that I customized a lot. And as such, gained the knowledge to fix any incoming incompatibilities. Most of the time, there is nothing to fix... Except on breaking changes, which are announced as breaking.

I am now using the addon for leaders b, q and w. You chose the most difficult leader to change... I would be curious to know the new purpose of your leader c. 

I do wish to say that as a user I am just expressing an opinion. If it were up to me, the second use case we where talking about would be supported by LazyVim itself.

> I can see a middle way where the keybinds (again a ExposedFunctionality -> Keybind map) are a separate part of the plugin configuration thus avoiding the problems of  incompatible configs.

Interesting!

Best regards!

**Comment by folke** [comment]
Thank you for the suggestion, but I think this would complicate things and would be too different from how you configure plugins with lazy.nvim.

And thank you @abeldekat to provide a plugin to make it possible to change top-level keys! 
That should solve op's issue.

**Comment by abeldekat** [comment]
You're welcome!

**Comment by abeldekat** [comment]
You're welcome!

**Comment by pathetic-lynx** [comment]
This plugin stopped working some time ago. I have been able to fix one problem in the e2e test (https://github.com/pathetic-lynx/lazyvim-menu-addon/commit/0813f0c16847a3bf92fd87cc04733c5540279107) but the other problem eludes me. I tried pinning LazyVim, lazy.nvim and which-key.nvim to older versions but to no avail.

Could @abeldekat or someone more familiar with the architecture look into this? I spent quite some time trying to fix this but I'm out of ideas :/
